app/src/authFetch.ts

```diff
@@ -0,0 +1,60 @@
+import { BASE_URL } from "@phoenix/config";
+
+const REFRESH_URL = BASE_URL + "/auth/refresh";
+
+class UnauthorizedError extends Error {
+  constructor() {
+    super("Unauthorized");
+  }
+}
+
+/**
+ * A wrapper around fetch that retries the request if the server returns a 401.
+ */
+export async function authFetch(
+  input: RequestInfo | URL,
+  init?: RequestInit
+): Promise<Response> {
+  try {
+    return await fetch(input, init).then((response) => {
+      if (response.status === 401) {
+        // If the server returns a 401, we should try to refresh the token
+        throw new UnauthorizedError();
+      }
+      return response;
+    });
+  } catch (error) {
+    if (error instanceof UnauthorizedError) {
+      // If the server returns a 401, we should try to refresh the token
+      await refreshTokens();
+      // Retry the original request
+      return fetch(input, init);
+    }
+  }
+  throw new Error("An unexpected error occurred while fetching data");
+}
+
+let refreshPromise: Promise<Response> | null = null;
+
+async function refreshTokens(url: string = REFRESH_URL): Promise<Response> {
+  if (refreshPromise) {
+    // There is already a refresh request in progress, so we should wait for it
+    return refreshPromise;
+  }
+  // This function should make a request to the server to refresh the access token
+  // eslint-disable-next-line no-console
+  console.log("Refreshing tokens");
+  refreshPromise = fetch(url, {
+    method: "POST",
+  }).then((response) => {
+    if (!response.ok) {
+      // for now force redirect to login page. This could re-throw with a custom error
+      // But for now, we'll just redirect
+      window.location.href = "/login";
+    }
+    // Clear the refreshPromise so that future requests will trigger a new refresh
+    refreshPromise = null;
+    return response;
+  });
+  return refreshPromise;
```

Comment:
does this need to return? doesn't seem to be used up above

Reply:
you can't await it otherwise. Can you clarify your question a bit?

Reply:
ahh right, sorry was thinking it was returning the response of the promise not the promise